go/ql/lib/semmle/go/security/JWT.qll

```diff
@@ -0,0 +1,281 @@
+import go
+
+/**
+ * A class that contains the following function and method:
+ *
+ * func (p *Parser) Parse(tokenString string, keyFunc Keyfunc)
+ *
+ * func Parse(tokenString string, keyFunc Keyfunc)
+ */
+class GolangJwtParse extends Function {
+  GolangJwtParse() {
+    exists(DataFlow::Function f |
+      f.hasQualifiedName([
+          "github.com/golang-jwt/jwt", "github.com/golang-jwt/jwt/v4",
+          "github.com/golang-jwt/jwt/v5", "github.com/dgrijalva/jwt-go",
+          "github.com/dgrijalva/jwt-go/v4",
+        ], "Parse")
+    |
+      this = f
+    )
+    or
+    exists(DataFlow::Method f |
+      f.hasQualifiedName([
+          "github.com/golang-jwt/jwt.Parser", "github.com/golang-jwt/jwt/v4.Parser",
+          "github.com/golang-jwt/jwt/v5.Parser", "github.com/dgrijalva/jwt-go.Parser",
+          "github.com/dgrijalva/jwt-go/v4.Parser"
+        ], "Parse")
+    |
+      this = f
+    )
+  }
+
+  int getKeyFuncArgNum() { result = 1 }
+
+  DataFlow::Node getKeyFuncArg() { result = this.getACall().getArgument(this.getKeyFuncArgNum()) }
+}
```

Comment:
A nicer way to do this is to define an abstract class, like this:
```suggestion
abstract class JwtParseFunction extends Function {
  abstract int getKeyFuncArgNum();

  DataFlow::Node getKeyFuncArg() { result = this.getACall().getArgument(this.getKeyFuncArgNum()) }
}

class GolangJwtParse extends JwtParseFunction {
  GolangJwtParse() {
    exists(string pkg |
      pkg =
        [
          "github.com/golang-jwt/jwt", "github.com/golang-jwt/jwt/v4",
          "github.com/golang-jwt/jwt/v5", "github.com/dgrijalva/jwt-go",
          "github.com/dgrijalva/jwt-go/v4",
        ]
    |
      this.hasQualifiedName(pkg, "Parse") or
      this.(Method).hasQualifiedName(pkg, "Parser", "Parse")
    )
  }

  override int getKeyFuncArgNum() { result = 1 }
}
```
This avoids duplicating lots of code. You should apply the same changes to the other classes in this file which implement the predicate `getKeyFuncArg()`. At the same time I have done some mild refactoring to avoid listing all the packages twice and to use the 3-argument version of `hasQualifiedName` for `Method`s, which is a bit easier to read.